lib/ansible/modules/tacp/tacp_instance.py

```diff
@@ -80,278 +78,713 @@
-        vtx_enabled=dict(type='bool', default=True, required=False),
-        auto_recovery_enabled=dict(type='bool', default=True, required=False),
-        description=dict(type='str', required=False),
-        vm_mode=dict(type='str', default='Enhanced', choices=['enhanced', 'Enhanced',
-                                                              'compatibility', 'Compatibility']),
-        application_group={
-            'type': 'str',
-            'required': False,
-        },
+RESULT = {
+    'changed': False,
+    'args': []
+}
+
+MODULE = AnsibleModule(
+    argument_spec=MODULE_ARGS,
+    supports_check_mode=True
+)
+
+# Define configuration
+CONFIGURATION = tacp_utils.get_configuration(MODULE.params['api_key'],
+                                             MODULE.params['portal_url'])
+API_CLIENT = tacp.ApiClient(CONFIGURATION)
+
+RESOURCES = {
+    'app': tacp_utils.ApplicationResource(API_CLIENT),
+    'app_group': tacp_utils.ApplicationGroupResource(API_CLIENT),
+    'datacenter': tacp_utils.DatacenterResource(API_CLIENT),
+    'migration_zone': tacp_utils.MigrationZoneResource(API_CLIENT),
+    'storage_pool': tacp_utils.StoragePoolResource(API_CLIENT),
+    'template': tacp_utils.TemplateResource(API_CLIENT),
+    'update_app': tacp_utils.UpdateApplicationResource(API_CLIENT),
+    'vlan': tacp_utils.VlanResource(API_CLIENT),
+    'vnet': tacp_utils.VnetResource(API_CLIENT)
+}
+
+
+def fail_with_reason(reason):
+    RESULT['msg'] = reason
+    MODULE.fail_json(**RESULT)
+
+
+def fail_and_rollback_instance_creation(reason, instance):
+    RESULT['msg'] = reason + "\n Rolled back application instance creation."
+    RESOURCES['app'].delete(instance.uuid)
+    MODULE.fail_json()
+
+
+def get_parameters_to_create_new_application(playbook_instance):
+    """Given the input configuration for an instance, generate
+        parameters for creating the appropriate payload elsewhere.
 
+    Args:
+        playbook_instance (dict): The specified instance configuration from
+        the playbook input
+
+    Returns:
+        dict: The parameters to be provided to an ApiCreateApplicationPayload
+            object.
+    """
+    parameters_to_create_new_application = {}
+    parameters_to_create_new_application['instance_name'] = playbook_instance['name']  # noqa
+
+    resources = [
+        'datacenter',
+        'migration_zone',
+        'storage_pool',
+        'template'
+    ]
+
+    for resource in resources:
+        resource_list = RESOURCES[resource].filter(
+            name=playbook_instance[resource])
+        if not resource_list:
+            fail_with_reason("{} resource {} does not exist, cannot continue.".format(  # noqa
+                resource, playbook_instance[resource]
+            ))
+        resource_uuid = resource_list[0].uuid
+
+        parameters_to_create_new_application['{}_uuid'.format(
+            resource)] = resource_uuid
+
+        if resource == 'template':
```

Comment:
Jsyk, its generally a good idea to prefer less indents 
```suggestion
        if resource != 'template':
            continue
        # do stuff if template
```